src/parse.cpp

```diff
@@ -73,11 +73,16 @@ parser::function_guard parser::enter_function(function_attributes attributes) {
                         was_in_switch_statement);
 }
 
+parser::depth_guard parser::enter_depth() {
+  return depth_guard(this, this->depth_);
+}
+
 parser::loop_guard parser::enter_loop() {
   return loop_guard(this, std::exchange(this->in_loop_statement_, true));
 }
 
 expression* parser::parse_expression(precedence prec) {
+  depth_guard guard = this->enter_depth();
```

Comment:
Nit: If we simplify `depth_guard`'s constructor, then we can avoid the `enter_depth` function:
```suggestion
  depth_guard guard(this);
```